### DISCUSSION TITLE
Client-side state system

### DISCUSSION THREAD

**Discussion by Nesqwik** [lenra-io/lenra_cli#97] [General]
The big issue with Lenra right now is the lack of client-side state. 
For now, if we want to control a popup for example, we have to link the popup "open" state to a mongo document.

We could create a client-side state to manage this kind of cases.

## My idea 
A `state` component could be created by the developer to create a state around one or more components.
This `state` component have a `name` that can be used anywhere to get the data.
 
Example : 
```json
{
    "type": "state",
    "name": "myState",
    "init": {
        "open": false
    },
    "child": {
        "type": "modal",
        "child": {...},
        "open": "#myState.open"
    }
}
```

Now to manipulate this state, there is multiple options.
- We could change it from a listener via an API (but not full client-side)
- We could add command to manipulate the state (for example in the "onPressed" object)

We can also add automatic client-side state like "hover", "focused" and so on. These are fully automatic and can only be read.

A (messy) example that need some work 
```json
{
    "type": "state",
    "name": "myState",
    "init": {
        "open": false
    },
    "child": {
        "type": "flex",
        "children": [{
            "type": "modal",
            "child": {...},
            "open": "#myState.flex.text.hover"
        }, {
            "type": "text",
            "value": "hover me !",
            "enableState": {
                "hover": true
            }
        }]
    }
}
```

**Comment by jonas-martinez** [comment]
This seems like a good idea to me. I'm doing some research to be sure that there is no big downside to this.
